### PR TITLE
return right blockHash in logs (fix #208)

### LINKED ISSUE #208: eth_getLogs returns wrong blockHash
Now the rpc node of xinfin blockchain returns wrong blockHash when call method `eth_getLogs`. Below is an example:

**request:**
```shell
curl -s -X POST https://arpc.apothem.network/ -H "Content-Type: application/json" -d '
{
    "id": 1610,
    "jsonrpc": "2.0",
    "method": "eth_getLogs",
    "params": [
        {
            "address": "0x53350795c11cee781a7e174479778f848d76ab2a",
            "fromBlock": "0x22b2277",
            "toBlock": "0x22b2277",
            "topics": [
                [
                    "0x8c5be1e5ebec7d5bd14f71427d1e84f3dd0314c0f7b2291e5b200ac8c7c3b925",
                    "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef",
                    "0x6a12b3df6cba4203bd7fd06b816789f87de8c594299aed5717ae070fac781bac"
                ]
            ]
        }
    ]
}' | jq
```

**response:**
```json
{
  "jsonrpc": "2.0",
  "id": 1610,
  "result": [
    {
      "address": "xdc53350795c11cee781a7e174479778f848d76ab2a",
      "topics": [
        "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef",
        "0x0000000000000000000000000000000000000000000000000000000000000000",
        "0x000000000000000000000000ce77393582d39048a4fc73770ab19d2b239c45dd"
      ],
      "data": "0x000000000000000000000000000000000000000000084595161401484a000000",
      "blockNumber": "0x22b2277",
      "transactionHash": "0xbebc0bab1a0f9844b86b8edf3a169bb1eef9337ac8d2684d5ae758b6521c12fe",
      "transactionIndex": "0x1",
      "blockHash": "0x194a513ea40c94fccaca470cccb24c2ff3bd6398dec9a04e3eb3a4bfef8a330e",
      "logIndex": "0x2",
      "removed": false
    }
  ]
}
```

The right blockHash shoud be `0xe8ec0bfc021f839fd0a69ae89689e309632af2cc434e8a096732068f0736864b` according to https://explorer.apothem.network/blocks/0x22b2277#transactions. The blockHash `0x194a513ea40c94fccaca470cccb24c2ff3bd6398dec9a04e3eb3a4bfef8a330e` in response is wrong.

### PR DESCRIPTION
**Pull request by gzliudan**
### Problem 1: wrong blochHash

Some JSONRPC methods return logs array in response, such as:

- `eth_getLogs`
- `eth_getFilterLogs`
- `eth_getFilterChanges`
- `eth_getTransactionReceipt`

But the blockHash filed in log object is wrong now, such as reported in #208. This bug cause some software which call `eth_getLogs` can not work normally, such as:

- [graph-node](https://github.com/graphprotocol/graph-node)
- [gnosis safe](https://safe.global/)

### Problem 2: unstable latest block

As described in issue #232, the latest block is replaced by final block sometimes. This issue makes blockHash wrong when the response includes the latest block also. Only method `eth_getTransactionReceipt` won't face this problem since it does not use block number in parameters. But the other three methods include block number in parameters, eg for the below requests:

```shell
curl -s -X POST http://103.101.129.240:8545/ -H "Content-Type: application/json" -d '
{
    "id": 12,
    "jsonrpc": "2.0",
    "method": "eth_getLogs",
    "params": [
        {
        }
    ]
}' | jq

curl -s -X POST http://103.101.129.240:8545/ -H "Content-Type: application/json" -d '
{
    "id": 213,
    "jsonrpc": "2.0",
    "method": "eth_newFilter",
    "params": [
        {
        }
    ]
}' | jq
```

The `fromBlock` and `toBlock` parameters in the above requests has default value "latest".

### 1. Replace wrong blockHash

This PR replace wrong blockHash with right values in the response for the above four JSONRPC methods to solve problem 1.

### 2. Use previous block instead of the latest block

This PR use previous block instead of the latest block for methods `eth_getLogs` and `eth_getFilterLogs` to solve problem 2. Assume the current latest block number is 1000 now, then this PR will set `fromBlock` and `toBlock` to 999, not 1000 for request.

### 3. Filter out the latest block

This PR checks each of log object's block number in the response for method `eth_getFilterChanges`. If find a log belongs to the latest block, then remove it from reponse, and keep it in buffer until next polling. It will be returned in next polling since the latest block number is increased at that time.

#### 1.1 without patch

return wrong block hash: 0x194a513ea40c94fccaca470cccb24c2ff3bd6398dec9a04e3eb3a4bfef8a330e

```shell
curl -s -X POST https://erpc.apothem.network/ -H "Content-Type: application/json" -d '
{
    "id": 1,
    "jsonrpc": "2.0",
    "method": "eth_getTransactionReceipt",
    "params": [
        "0xbebc0bab1a0f9844b86b8edf3a169bb1eef9337ac8d2684d5ae758b6521c12fe"
    ]
}' | jq
```

---

#### 1.2 with patch

return right block hash: 0xe8ec0bfc021f839fd0a69ae89689e309632af2cc434e8a096732068f0736864b

```shell
curl -s -X POST http://103.101.129.240:8545/ -H "Content-Type: application/json" -d '
{
    "id": 2,
    "jsonrpc": "2.0",
    "method": "eth_getTransactionReceipt",
    "params": [
        "0xbebc0bab1a0f9844b86b8edf3a169bb1eef9337ac8d2684d5ae758b6521c12fe"
    ]
}' | jq
```

#### 2.1 without patch

return wrong block hash: 0x851cb67720d6a78bef95be0148d0ef620c6daf6fd662befcccf38097f2973873

```shell
curl -s -X POST https://erpc.apothem.network/ -H "Content-Type: application/json" -d '
{
    "id": 3,
    "jsonrpc": "2.0",
    "method": "eth_getTransactionReceipt",
    "params": [
        "0x0b1c97fabd7d7a9505e5df426ecf5f8ebbf8bae988eb64e394c177d656cb7199"
    ]
}' | jq
```

---

#### 2.2 with patch

return right block hash: 0xd10ca14840489e88070508428fed14f96e39ad20cc524468a0d54a221b9b87e1

```shell
curl -s -X POST http://103.101.129.240:8545/ -H "Content-Type: application/json" -d '
{
    "id": 4,
    "jsonrpc": "2.0",
    "method": "eth_getTransactionReceipt",
    "params": [
        "0x0b1c97fabd7d7a9505e5df426ecf5f8ebbf8bae988eb64e394c177d656cb7199"
    ]
}' | jq
```

#### 1.1 without patch

return wrong block hash: 0x194a513ea40c94fccaca470cccb24c2ff3bd6398dec9a04e3eb3a4bfef8a330e

```shell
curl -s -X POST https://arpc.apothem.network/ -H "Content-Type: application/json" -d '
{
    "id": 11,
    "jsonrpc": "2.0",
    "method": "eth_getLogs",
    "params": [
        {
            "fromBlock": "0x22b2277",
            "toBlock": "0x22b2277"
        }
    ]
}' | jq
```

---

#### 1.2 with patch

return right block hash: 0xe8ec0bfc021f839fd0a69ae89689e309632af2cc434e8a096732068f0736864b

```shell
curl -s -X POST http://103.101.129.240:8545/ -H "Content-Type: application/json" -d '
{
    "id": 12,
    "jsonrpc": "2.0",
    "method": "eth_getLogs",
    "params": [
        {
            "fromBlock": "0x22b2277",
            "toBlock": "0x22b2277"
        }
    ]
}' | jq
```

#### 2.1 without patch

return wrong block hash:

- [45631260](https://explorer.apothem.network/blocks/45631260#transactions): 0x851cb67720d6a78bef95be0148d0ef620c6daf6fd662befcccf38097f2973873
- [45631261](https://explorer.apothem.network/blocks/45631261#transactions): 0x2a24c1d6b6c8a08fc805022b82f1b5e91851b92476a5ed0bba174ebbcbca45a1
- [45631262](https://explorer.apothem.network/blocks/45631262#transactions): 0xfa2059b48d7df89b1c8f366953f56e87c778808e85997a9a2b0e634c7037d470

```shell
curl -s -X POST https://erpc.apothem.network/ -H "Content-Type: application/json" -d '
{
    "id": 13,
    "jsonrpc": "2.0",
    "method": "eth_getLogs",
    "params": [
        {
            "fromBlock": "0x2b8471c",
            "toBlock": "0x2b8471e"
        }
    ]
}' | jq | grep -i block
```

---

#### 2.2 with patch

return right block hash:

- [45631260](https://explorer.apothem.network/blocks/45631260#transactions): 0xd10ca14840489e88070508428fed14f96e39ad20cc524468a0d54a221b9b87e1
- [45631261](https://explorer.apothem.network/blocks/45631261#transactions): 0x781d947687cdad5cb1b0022c13effc9904b70a37bf1af0afe425ccd287225794
- [45631262](https://explorer.apothem.network/blocks/45631262#transactions): 0xb6a64ba4d6ffefd8040c1fdaf3a75e6612a7346062b48a6e0ece68d57eb3f25f

```shell
curl -s -X POST http://103.101.129.240:8545/ -H "Content-Type: application/json" -d '
{
    "id": 14,
    "jsonrpc": "2.0",
    "method": "eth_getLogs",
    "params": [
        {
            "fromBlock": "0x2b8471c",
            "toBlock": "0x2b8471e"
        }
    ]
}' | jq | grep -i block
```

#### 1.1 without patch

return wrong block hash:

```shell
# create a filter
curl -s -X POST https://erpc.apothem.network/ -H "Content-Type: application/json" -d '
{
    "id": 111,
    "jsonrpc": "2.0",
    "method": "eth_newFilter",
    "params":[
        {
            "fromBlock": "0x2bb7231",
            "toBlock": "0x2bb7233"
        }
    ]
}' | jq

# 0x6e12f3f9faae153addcac28b33ede959 is filter id in above response, replace it with your actual value
FILTER_ID="0xaf54be244f6c1e2b927324f1882ebf46"

curl -s -X POST https://erpc.apothem.network/ -H "Content-Type: application/json" -d '
{
    "id": 112,
    "method": "eth_getFilterLogs",
    "params": ["'"${FILTER_ID}"'"]
}' | jq | grep block
```

| block number                                                              | right block hash                                                   | return wrong block hash                                            |
| :------------------------------------------------------------------------ | :----------------------------------------------------------------- | :----------------------------------------------------------------- |
| [45838897](https://explorer.apothem.network/blocks/45838897#transactions) | 0x1a7ed2982008d515dfe6f9b779be673268d9ec84d95cf254963b8c8c0f496dee | 0x0e3bef8f80d6194dd6da777e6d275b8632d4c8ec2bd8396e75159effebd9cd2b |
| [45838898](https://explorer.apothem.network/blocks/45838898#transactions) | 0x24b3018c78198ba2e4eb86b4a02eb5e06ea9230974b676c241ccc516ba8caf3c | 0x55ecfa357b1d699294dcf9bc1ea935d5aa44a771011964f3470efe976944b8ac |
| [45838899](https://explorer.apothem.network/blocks/45838899#transactions) | 0x00743071dc87ea3de489b53218b12d0c39fea13a3192a67c521556743d60b11b | 0x4dea4cda26b661c4d17fdf496a5e72be48ada092a789979d29b8498af8e64c32 |

---

#### 1.2 with patch

return right block hash:

```shell
# create a filter
curl -s -X POST http://103.101.129.240:8545/ -H "Content-Type: application/json" -d '
{
    "id": 113,
    "jsonrpc": "2.0",
    "method": "eth_newFilter",
    "params": [
        {
            "fromBlock": "0x2bb7231",
            "toBlock": "0x2bb7233"
        }
    ]
}' | jq

# 0x810ef8bf7733dd95ef52174b7b7ed676 is filter id in above response, replace it with your actual value
FILTER_ID="0x810ef8bf7733dd95ef52174b7b7ed676"

curl -s -X POST http://103.101.129.240:8545/ -H "Content-Type: application/json" -d '
{
    "id": 114,
    "method": "eth_getFilterLogs",
    "params": ["'"${FILTER_ID}"'"]
}' | jq | grep block
```

| block number                                                              | right block hash                                                   |
| :------------------------------------------------------------------------ | :----------------------------------------------------------------- |
| [45838897](https://explorer.apothem.network/blocks/45838897#transactions) | 0x1a7ed2982008d515dfe6f9b779be673268d9ec84d95cf254963b8c8c0f496dee |
| [45838898](https://explorer.apothem.network/blocks/45838898#transactions) | 0x24b3018c78198ba2e4eb86b4a02eb5e06ea9230974b676c241ccc516ba8caf3c |
| [45838899](https://explorer.apothem.network/blocks/45838899#transactions) | 0x00743071dc87ea3de489b53218b12d0c39fea13a3192a67c521556743d60b11b |

#### 2.1 without patch

return wrong block hash:

```shell
# create a filter
curl -s -X POST https://erpc.apothem.network/ -H "Content-Type: application/json" -d '
{
    "id": 211,
    "jsonrpc": "2.0",
    "method": "eth_newFilter",
    "params":[
        {
        }
    ]
}' | jq

# 0x89a75daaa6733b9887172617aafe534d is filter id in above response, replace it with your actual value
FILTER_ID="0x89a75daaa6733b9887172617aafe534d"

curl -s -X POST https://erpc.apothem.network/ -H "Content-Type: application/json" -d '
{
    "id": 212,
    "method": "eth_getFilterLogs",
    "params": ["'"${FILTER_ID}"'"]
}' | jq | grep block

# 0x2be22ee is the block number in above response, replace it with your actual value
BLOCK_NUMBER="0x2be22ee"

# fetch block hash by method eth_getBlockByNumber
curl -s -X POST https://erpc.apothem.network/ -H "Content-Type: application/json" -d '
{
    "id": 21,
    "jsonrpc": "2.0",
    "method": "eth_getBlockByNumber",
    "params":["'"${BLOCK_NUMBER}"'", false]
}' | jq | grep hash
```

This right block hash of [block number 46015214]((https://explorer.apothem.network/blocks/46015214#transactions) should be: 0x5c14e380d1845a6358547d44cc68ed7647d2e5b0916415d85a00f61b9b0ebf9c

---

#### 2.2 with patch

return right block hash:

```shell
# create a filter
curl -s -X POST http://103.101.129.240:8545/ -H "Content-Type: application/json" -d '
{
    "id": 213,
    "jsonrpc": "2.0",
    "method": "eth_newFilter",
    "params": [
        {
        }
    ]
}' | jq

# 0x14c4d591470e8dbf96d05280d0a2fca2 is filter id in above response, replace it with your actual value
FILTER_ID="0x14c4d591470e8dbf96d05280d0a2fca2"

curl -s -X POST http://103.101.129.240:8545/ -H "Content-Type: application/json" -d '
{
    "id": 214,
    "method": "eth_getFilterLogs",
    "params": ["'"${FILTER_ID}"'"]
}' | jq | grep block

# 0x2be23f4 is the block number in above response, replace it with your actual value
BLOCK_NUMBER="0x2be23f4"

# fetch block hash by method eth_getBlockByNumber
curl -s -X POST http://103.101.129.240:8545/ -H "Content-Type: application/json" -d '
{
    "id": 21,
    "jsonrpc": "2.0",
    "method": "eth_getBlockByNumber",
    "params":["'"${BLOCK_NUMBER}"'", false]
}' | jq | grep hash
```

This right block hash of [block number 46015476]((https://explorer.apothem.network/blocks/46015476#transactions) should be: 0x6827bf61c17d642717139f6bb6d533ac339865c1fb9536d730dacc6fefe25a6b

### without patch

```shell
# create a filter
curl -s -X POST https://erpc.apothem.network/ -H "Content-Type: application/json" -d '
{
    "id": 1111,
    "jsonrpc": "2.0",
    "method": "eth_newFilter",
    "params": [
        {}
    ]
}' | jq

# 0x68ce60ffdb0c9480c307b0c3d2ae9391 is filter id in above response, replace it with your actual value
FILTER_ID="0x68ce60ffdb0c9480c307b0c3d2ae9391"

curl -s -X POST https://erpc.apothem.network/ -H "Content-Type: application/json" -d '
{
    "id": 1112,
    "method": "eth_getFilterChanges",
    "params": ["'"${FILTER_ID}"'"]
}' | jq | grep block
```

| block number                                                              | right block hash                                                   | return wrong block hash                                            |
| :------------------------------------------------------------------------ | :----------------------------------------------------------------- | :----------------------------------------------------------------- |
| [46012631](https://explorer.apothem.network/blocks/46012631#transactions) | 0xc9c375a8758d9b50c5292b714d99b8372ab45b9ec138d7187d31be0cb4ce8293 | 0xd288fc1f08a68188d31879f91b43bb640cbbe64248bea8956f7c2e188d0f9b35 |
| [46012632](https://explorer.apothem.network/blocks/46012632#transactions) | 0x4519feb874b65166614ad9051f3bfec2119c1e7c5a17a66cc6b34263b7150201 | 0x5699860a5f43dbb8df294a0d289a8e92921977e447562aaea96b2e9146b827fe |
| [46012633](https://explorer.apothem.network/blocks/46012633#transactions) | 0x98ca3c24ca35bbea26a646b0c5044d0e417f644ad4859e569d7c227d5f8322e1 | 0x5e453ed091a11ccb8472d297e315dc24a14cd46a59f95d330f86b6e8ddf85b5a |
| [46012634](https://explorer.apothem.network/blocks/46012634#transactions) | 0x4d73756f56dfbe96897e5fca1db730dcadd712d0c24e23acc738ccfeb50f610e | 0xc02434f03580ec7b7477537df81f009aa8474043a6b947c32bacb4614d5aac0f |
| [46012635](https://explorer.apothem.network/blocks/46012635#transactions) | 0x6460dc32d700bb9af70537a4b04af0c6dd088b741fdc5bbab12a2bf3b08ea678 | 0x31883a10bcdfa46cd4802f7358fe6791059963451e9aba2095c29b3d1ebc81b5 |
| [46012636](https://explorer.apothem.network/blocks/46012636#transactions) | 0xbe7584e54ed359814f38453eee00a029bc3136a30b005e448e0f078f6b1ad4dc | 0x34ab5bee1cfded4a7bb0b3f221f6e576c42aa115d0c89a47f37ccf90806e2f2c |
| [46012637](https://explorer.apothem.network/blocks/46012637#transactions) | 0xc73c19ba416affd8c516c4781008a5630809abfaaecd831a8b7bef8cf66bec65 | 0xe9371b3c728e110eb1a6288b27cdd9035b66d9bec470a0bd5dcfa009e4c09282 |
| [46012638](https://explorer.apothem.network/blocks/46012638#transactions) | 0xb0a80f91ccd7d4965fe88b234c62629c2c14ed1ebdb331ab8acdfdf8e82647dc | 0x632c8f424752d6b365f27e8c6131f9c92849284df070050a692750841c78b4e7 |
| [46012639](https://explorer.apothem.network/blocks/46012639#transactions) | 0x9a37917fefa343b9804d6d99ffa054a1428167cdce6ba067657a65b4d1902862 | 0xcc7220f6363610b4d66ce34d8218b52947d678e05a408d8574349febf3f4c852 |

### with patch

```shell
# create a filter
curl -s -X POST http://103.101.129.240:8545/ -H "Content-Type: application/json" -d '
{
    "id": 1113,
    "jsonrpc": "2.0",
    "method": "eth_newFilter",
    "params": [
        {}
    ]
}' | jq

# 0x8936ced335de251e1211799d7550b7ca is filter id in above response, replace it with your actual value
FILTER_ID="0x8936ced335de251e1211799d7550b7ca"

curl -s -X POST http://103.101.129.240:8545/  -H "Content-Type: application/json" -d '
{
    "id": 1114,
    "method": "eth_getFilterChanges",
    "params": ["'"${FILTER_ID}"'"]
}' | jq | grep block
```

|                               block number                                |                          right block hash                          |
| :-----------------------------------------------------------------------: | :----------------------------------------------------------------: |
| [46013738](https://explorer.apothem.network/blocks/46013738#transactions) | 0xe2e20073ab2eb0cb7ee76fd744e01a90e9c15ab14473f577429e69bea2635812 |
| [46013739](https://explorer.apothem.network/blocks/46013739#transactions) | 0xd681e4a7e8d0d74a6ab235b5adebdf4b187fe52284a91531fb223f9adf1913f8 |
| [46013740](https://explorer.apothem.network/blocks/46013740#transactions) | 0x412062ec3a890e0f3b4bda0cad82804f63cc2c3383bea1e873eb904978522ade |

### CODE REVIEW

**Review by liam-lai — APPROVED** [review]

File: eth/filters/api.go [hunk 1]
```diff
@@ -416,8 +417,23 @@ func (api *PublicFilterAPI) GetFilterChanges(id rpc.ID) (interface{}, error) {
 			f.hashes = nil
 			return returnHashes(hashes), nil
 		case LogsSubscription:
-			logs := f.logs
-			f.logs = nil
+			// not reeturn the log in the lastest block to bypass issue #232
```
Comment (liam-lai) [inline eth/filters/api.go:420]: return

Comment (gzliudan) [reply eth/filters/api.go:420]: Yes, this is typo. I correct it at once.

Comment (gzliudan) [reply eth/filters/api.go]: I corrected this typo just now.

**Comment by gzliudan** [comment]
Deprecated by #320

**Pull request closed without merging**